### PR TITLE
RavenDB-19529: Missing compression dictionary (OpenTable bug fix)

### PR DESCRIPTION
**Pull request by ArieSLV**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19529/Missing-compression-dictionary

### Additional description

The issue arose while performing a merged write transaction. 
First, a table for a specific collection was opened. 
After that, a document that met the following criteria was updated using a `PUT` command:
1. The document should contain random data that is difficult to compress.
2. After compression, the compressed buffer should not fit into the small section.
3. The size of the `compressed buffer` + `sizeof(RawDataSection.RawDataEntrySizes)` (4 bytes) should be greater than `RawDataSection.MaxItemSize` (4064 bytes).

The crux of the problem is that upon the first opening of the table, we cache the schema for the collection and use this schema for subsequent openings of this table.

For some operations, when opening the table, we were forcibly using a non-compressed schema, which ended up in the cache. During the next table opening, we took the non-compressed schema, which is fatal for the `Update` case when the `TableValue` fits into the large section. In this case, no attempt is made to perform `builder.TryCompression` because the schema is mistakenly marked as non-compressed. The `TableValueBuilder` doesn't get marked with the `Compressed` property set to `true`, and it ends up on the same page that was earlier flagged as compressed.

As a result, during the subsequent attempt to read such a `TableValue`, we look at the flag on the page and try to decompress. To do this, we read the number of the compression dictionary, receive a _garbage_ number, and either cannot find such a number or get the message `Unknown frame descriptor on Dictionary`.

### Type of change

- Bug fix

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 1]
```diff
@@ -1033,7 +1019,7 @@ public IEnumerable<Document> GetDocumentsFrom(DocumentsOperationContext context,
             if (collectionName == null)
                 yield break;
 
-            var table = context.Transaction.InnerTransaction.OpenTable(DocsSchema,
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/DocumentsStorage.cs:1036]: As I understand the issue was that we passed "blindly" `DocsSchema` without checking if a collection is compressed or not, is that correct? 

What will happen in a scenario where a collection was compressed and then the compression was disabled for it. Then we'll pass `DocsSchema`. How is that case supposed to work?

File: src/Raven.Server/Documents/DocumentDatabase.cs [hunk 2]
```diff
@@ -1680,6 +1683,27 @@ public DatabaseRecord ReadDatabaseRecord()
...
         }
 
+        public TableSchema GetDocsSchemaForCollection(CollectionName collection) =>
+            _documentsCompression.CompressAllCollections || _compressedCollections.Contains(collection.Name)
+                ? DocumentsStorage.CompressedDocsSchema
+                : DocumentsStorage.DocsSchema;
+
+        private void UpdateCompressionConfigurationFromDatabaseRecord(DatabaseRecord record)
+        {
+            if (_documentsCompression.Equals(record.DocumentsCompression))
```
Comment (grisha-kotler) [inline src/Raven.Server/Documents/DocumentDatabase.cs:1693]: What happens if the configuration changes while there is a running write transaction?
- The same test but starts without compression.
- After deleting the document, we update the compression configuration.
- The update will have the compressed schema while we take the table (from the cache) with the non-compressed schema.

This was also the previous behavior - let's open an issue for that.

Comment (ayende) [reply src/Raven.Server/Documents/DocumentDatabase.cs:1693]: It's fine to change the compression definition, we handle reading that on a per document basis, after all.
And worst case, we have a document that is compressed or not compressed when it should, but that is just a race condition without an impact.

File: src/Raven.Server/Documents/DocumentsStorage.cs (hunk 1 above)
Comment (ayende) [reply src/Raven.Server/Documents/DocumentsStorage.cs:1036]: For reads, this doesn't matter, we do the actual operation on a document by document scenario.
We can revert this safely.

File: src/Raven.Server/Documents/DocumentDatabase.cs (hunk 2 above)
Comment (grisha-kotler) [reply src/Raven.Server/Documents/DocumentDatabase.cs:1693]: I'm talking about changing it in the **middle** of the transaction.
1. We create 2 documents in a compressed collection.
2. Disable compression.
3. In the same transaction, delete one document and update the second one.
4. Inside the transaction, after we delete the document here:
https://github.com/ravendb/ravendb/pull/17659/files#diff-624ee6c1d1af1fe87f3d169216fa6ea981b9835799020d41b1625cb7bb450ec5R2186
Enable compression.
We already cached the table with non-compressed schema, and now we update the second document when the compression is enabled (which brings us to the same issue that we have in the test).

File: src/Raven.Server/Documents/DocumentsStorage.cs (hunk 1 above)
Comment (arekpalinski) [reply src/Raven.Server/Documents/DocumentsStorage.cs:1036]: @ArieSLV ?

Comment (ArieSLV) [reply src/Raven.Server/Documents/DocumentsStorage.cs:1036]: > As I understand the issue was that we passed "blindly" DocsSchema without checking if a collection is compressed or not, is that correct?

Correct

> What will happen in a scenario where a collection was compressed and then the compression was disabled for it. Then we'll pass DocsSchema. How is that case supposed to work?

As Oren previously mentioned, the schema is important at the time of writing. For entries placed in a small section, we set the flag Compressed per entry: https://github.com/ravendb/ravendb/blob/1bae88adfcc53300d50e36561320c1e53a3246e0/src/Voron/Data/RawData/RawDataSection.cs#L229C24-L229C24

In cases where an entry does not fit in the small section, we use a large one and apply the Compressed flag to the page.

During reading, we use these flags regardless of the current schema settings.

Comment (arekpalinski) [reply src/Raven.Server/Documents/DocumentsStorage.cs:1036]: Thank you. It's clear to me now.

**Review by arekpalinski — APPROVED** [review]

**Pull request merged**